pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/XPathRuleTest.java

```diff
@@ -92,13 +93,30 @@ public void testFnPrefixOnSaxon() throws Exception {
     }
 
     @Test
-    public void testNoFnPrefixOnSaxon() throws Exception {
+    public void testNoFnPrefixOnSaxon() {
         XPathRule rule = makeXPath("//VariableDeclaratorId[matches(@Name, 'fiddle')]");
         Report report = getReportForTestString(rule, TEST2);
         RuleViolation rv = report.getViolations().get(0);
         assertEquals(3, rv.getBeginLine());
     }
 
+    @Test
+    public void testSimpleQueryIsRuleChain() {
+        // ((/)/descendant::element(Q{}VariableDeclaratorId))[matches(convertUntyped(data(@Name)), "fiddle", "")]
+        assertIsRuleChain("//VariableDeclaratorId[matches(@Name, 'fiddle')]");
+    }
+
+    @Test
+    public void testSimpleQueryIsRuleChain2() {
+        // docOrder(((/)/descendant-or-self::node())/(child::element(ClassOrInterfaceType)[typeIs("java.util.Vector")]))
```

Comment:
This is the "inconsistent parsing" I'm referring to. The comments reflect how Saxon parses the expressions here. Those two rules are seemingly very similar, yet saxon parses the above as `(root/descendant::...)[...filter...]` and this one as `(root/descendant-or-self::node())/(child:...[...filter...])`. The solution is to pull up the filter expression to the top-level (like it is in the above expression), then reduce the `root/descendant-or-self/child::...` to `root/descendant::...`. This matches the pattern that the RuleChainAnalyser recognizes.